hax/hax/motr/__init__.py

```diff
@@ -355,19 +357,31 @@ def ha_obj_state(st):
                                                             kv_cache=kv_cache)
         if not notes:
             return []
+        notes_pending = notes
+        notes_len_total = len(notes)
+        idx = 0
         message_ids: List[MessageId] = []
-        LOG.debug('broadcast ha states phase %s', broadcast_hax_only)
-        if broadcast_hax_only:
-            hax_endpoint = self.consul_util.get_hax_endpoint()
-            message_ids = self._ffi.ha_broadcast_hax_only(
-                self._ha_ctx, make_array(HaNoteStruct, notes), len(notes),
-                make_c_str(hax_endpoint))
-        else:
-            message_ids = self._ffi.ha_broadcast(
-                self._ha_ctx, make_array(HaNoteStruct, notes), len(notes))
-        LOG.debug(
-            'Broadcast HA state complete with the following message_ids = %s',
-            message_ids)
+        while notes_pending:
+            if len(notes_pending) > MAX_MOTR_NVEC_UPDATE_SZ:
+                notes_to_send = notes[idx:(idx + MAX_MOTR_NVEC_UPDATE_SZ)]
```

Comment:
[Optional]
Effectively `notes_pending` is used for `if len(notes_pending)` expression only while `notes_to_send` is constructed from the whole `notes` list instead by some reason. And because of that you also have to maintain `idx` counter and perform extra list copying (by means of slice operation).

The idea could be simpler: you need to split the overall `notes` into pages of 1024 items. Moreover, every moment of time you're interested in the first page only. So the algorithm could be simpler:
1. Cut first 1024 elements from the top of `notes` and send them
2. Save the rest of the elements in `notes`
3. Repeat while `notes` is not empty

Note: If this logic was refactored into a separate method or a class, it would be possible to cover it with unit tests easily.

Reply:
Yeah, I was refraining from modifying notes to maintain an original copy. I will try to move it out to another function.